lib/Sema/CSDiagnostics.h

```diff
@@ -3111,12 +3111,13 @@ class InvalidMemberReferenceWithinInitAccessor final
 /// \endcode
 class ConcreteTypeSpecialization final : public FailureDiagnostic {
   Type ConcreteType;
+  ValueDecl *Decl;
```

Comment:
I think actually a `PointerUnion<TypeBase *, ValueDecl *>` since like either or based on the `::diagnoseAsError()` implementation.

Reply:
True. There isn't actually a need for the `ValueDecl` in generating the diagnosis message either, so changed this to check whether the decl is a generic function up front, and then only store the ConcreteType again, and if it is nullptr, that means the other diagnosis.